### PR TITLE
TBOX-155 Rename CSV columns

### PR DESCRIPTION
**Pull request by ErinCompaan**
# ↪️ Pull Request

Add optional parameter to csv export to allow for renaming columns. 

Parameter is a dictionary which maps from original (Tamr dataset) names to the new column names desired in the csv.  

CSV export will raise a ValueError if the supplied renaming dictionary would yield duplicate column names, and log a warning if a supplied `original_name` is not a heading in the dataset. 

## ✔️ PR Todo

- [x] Add documentation
- [x] Add examples
- [x] Ensure you are following the practices in the [contributor guide](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit) and [coding_standards](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards)
  * [Use consistent and descriptive name conventions](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.477s36w4rds9) 
  * [Uses google-style docstrings](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Google-style-docstrings)
  * [Uses versioned API/client methods only](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.yv6df99fyrq2)
  * [Avoid global variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-global-variables)
  * [Avoid giant try/catch](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-giant-try-/-except)
  * [Don't use mutable default variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Don%E2%80%99t-use-mutable-default-arguments)
  * [Use a `main()` function in your scripts](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Use-a-%E2%80%9Cmain()%E2%80%9D--function-in-your-scripts)
- [x] Added/updated testing for this change (ensure you are testing any changes/new code!)
- [x] Update title to link to TBOX issue (i.e. start pull request title with TBOX-xx) 
- [x] Approval from first code-reviewer
- [ ] Approval from second code-reviewer (only required if a second reviewer has left comments)
- [x] Pass all checks

### CODE REVIEW

File: tamr_toolbox/data_io/csv.py [hunk 1]
```diff
@@ -23,6 +23,7 @@ def from_dataset(
     *,
     csv_delimiter: str = ",",
     columns: Optional[List[str]] = None,
+    column_name_dict: Optional[dict] = None,
```
Comment (skalish) [inline tamr_toolbox/data_io/csv.py]: More specific type hint
```suggestion
    column_name_dict: Optional[Dict[str, str]] = None,
```

Comment (skalish) [reply tamr_toolbox/data_io/csv.py]: Feel free to improve type hints in this way on existing code in the files that you're touching as well.

File: tamr_toolbox/data_io/common.py [hunk 2]
```diff
@@ -127,3 +127,57 @@ def _check_columns_subset(
...
+        if not is_subset:
+            message = (
+                f"Some key columns in the column mapping {column_name_dict}"
+                "do not appear in specified Tamr dataset."
+            )
+            LOGGER.warning(message)
+
+        # Make dict mapping all attributes to themselves, and update with any renaming dict entries
+        column_name_dict = {k: v for k, v in column_name_dict.items() if k in columns}
+        full_column_name_dict = {**dict(zip(columns, columns)), **column_name_dict}
```
Comment (skalish) [inline tamr_toolbox/data_io/common.py]: A possible alternate statement (not necessary to change)
```suggestion
        full_column_name_dict = {column: column_name_dict.get(column, column) for k in columns}
```

File: tamr_toolbox/data_io/common.py [hunk 3]
```diff
@@ -127,3 +127,56 @@ def _check_columns_subset(
...
+
+        if not is_subset:
+            message = (
+                f"Some key column(s) in the column mapping {column_name_dict}"
+                "do not appear in specified Tamr dataset."
+            )
+            LOGGER.warning(message)
+
+        # Make dict mapping all attributes to themselves, and update with any renaming dict entries
+        full_column_name_dict = {col: column_name_dict.get(col, col) for col in columns}
```
Comment (skalish) [inline tamr_toolbox/data_io/common.py:172]: This line ensures that the order of the entries in `full_column_name_dict` matches the order of the entries in `columns` as desired.  Can you add a comment stating this so it's clear that this is happening (and is intentional)?

File: tests/data_io/test_csv.py [hunk 4]
```diff
@@ -284,3 +284,75 @@ def test_bad_encoding():
...
+            {"first_name": "new_first_name", "last_name": "family_name"},
+        ),  # basic test renaming test
+        (
+            None,
+            {"not_real_col": "full_name", "last_name": "family_name"},
+        ),  # test with column name not present in the data
+        (
+            ["tamr_id", "all_names", "ssn", "last_name", "first_name"],
+            {"last_name": "family_name"},
+        ),  # test with columns subset specified
```
Comment (skalish) [inline tests/data_io/test_csv.py]: Can you add a case where there are multiple columns renamed, and the order in `columns` and `column_name_dict` conflict?  The result should be that the order in `columns` is obeyed.

**Review by skalish — CHANGES_REQUESTED** [review]
This looks great!  I think another test case would be ideal, but then it should be ready for approval.

**Comment by skalish** [comment]
As more of an experiment than a requirement, can you locally run `pytest --cov=tamr_toolbox --cov-report=term-missing tests/` and see if any part of your contribution is not covered by tests?

**Comment by ErinCompaan** [comment]
> As more of an experiment than a requirement, can you locally run `pytest --cov=tamr_toolbox --cov-report=term-missing tests/` and see if any part of your contribution is not covered by tests?

Everything new is covered by tests:
```
tamr_toolbox/data_io/common.py                       59      0   100%
tamr_toolbox/data_io/csv.py                         113     24    79%   84-89, 94-99, 110-120, 178, 253, 265-268, 294-301
```

**Review by skalish — APPROVED** [review]
LGTM

**Pull request merged**